drivers/gpio/gpio_tca9538.c

```diff
@@ -0,0 +1,524 @@
+
+	/* Read Input Register */
+	uint8_t cmd = TCA9538_INPUT_PORT;
+
+	rc = i2c_write_read(cfg->i2c_dev, cfg->i2c_addr, &cmd,
+			    sizeof(cmd), &input_pin_data,
+			    sizeof(input_pin_data));
+	LOG_DBG("read %x got %d", input_pin_data, rc);
+
+	if (rc == 0) {
+		drv_data->pin_state.input = input_pin_data;
+		*value = (gpio_port_value_t)(drv_data->pin_state.input);
+	}
+
+	k_sem_give(&drv_data->lock);
+	return rc;
+}
+
+static int gpio_tca9538_port_write(const struct device *dev,
+				   gpio_port_pins_t mask,
+				   gpio_port_value_t value,
+				   gpio_port_value_t toggle)
+{
+	const struct tca9538_config *cfg = dev->config;
+	struct tca9538_drv_data *drv_data = dev->data;
+	uint8_t *outp = &drv_data->pin_state.output;
+	int rc;
+	uint8_t orig_out;
+	uint8_t out;
+
+	/* Can't do I2C bus operations from an ISR */
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	orig_out = *outp;
+	out = ((orig_out & ~mask) | (value & mask)) ^ toggle;
+
+	rc = i2c_reg_write_byte(cfg->i2c_dev, cfg->i2c_addr,
+				TCA9538_OUTPUT_PORT, out);
+
+	if (rc == 0) {
+		*outp = out;
+	}
+
+	k_sem_give(&drv_data->lock);
+
+	LOG_DBG("write %x msk %08x val %08x => %x: %d", orig_out, mask,
+		value, out, rc);
+
+	return rc;
+}
+
+static int gpio_tca9538_port_set_masked(const struct device *dev,
+					gpio_port_pins_t mask,
+					gpio_port_value_t value)
+{
+	return gpio_tca9538_port_write(dev, mask, value, 0);
+}
+
+static int gpio_tca9538_port_set_bits(const struct device *dev,
+				      gpio_port_pins_t pins)
+{
+	return gpio_tca9538_port_write(dev, pins, pins, 0);
+}
+
+static int gpio_tca9538_port_clear_bits(const struct device *dev,
+					gpio_port_pins_t pins)
+{
+	return gpio_tca9538_port_write(dev, pins, 0, 0);
+}
+
+static int gpio_tca9538_port_toggle_bits(const struct device *dev,
+					 gpio_port_pins_t pins)
+{
+	return gpio_tca9538_port_write(dev, 0, 0, pins);
+}
+
+static int gpio_tca9538_pin_interrupt_configure(const struct device *dev,
+						gpio_pin_t pin,
+						enum gpio_int_mode mode,
+						enum gpio_int_trig trig)
+{
+	const struct tca9538_config *cfg = dev->config;
+	struct tca9538_drv_data *drv_data = dev->data;
+
+	if (cfg->interrupt_enabled) {
+		/* Device does not support level-triggered interrupts. */
+		if (mode == GPIO_INT_MODE_LEVEL) {
+			return -ENOTSUP;
+		}
+
+		struct tca9538_irq_state *irq = &drv_data->irq_state;
+
+		k_sem_take(&drv_data->lock, K_FOREVER);
+
+		if (mode == GPIO_INT_MODE_DISABLED) {
+			irq->falling &= ~BIT(pin);
+			irq->rising &= ~BIT(pin);
+		} else { /* GPIO_INT_MODE_EDGE */
+			if (trig == GPIO_INT_TRIG_BOTH) {
+				irq->falling |= BIT(pin);
+				irq->rising |= BIT(pin);
+			} else if (trig == GPIO_INT_TRIG_LOW) {
+				irq->falling |= BIT(pin);
+				irq->rising &= ~BIT(pin);
+			} else if (trig == GPIO_INT_TRIG_HIGH) {
+				irq->falling &= ~BIT(pin);
+				irq->rising |= BIT(pin);
+			}
+		}
+		k_sem_give(&drv_data->lock);
+	} else {
+		return -ENOTSUP;
```

Comment:
In Zephyr we generally prefer early exit strategy (it's not a hard rule). To follow the convention you would need to 
```
	if (!cfg->interrupt_enabled) {
		return -ENOTSUP;
	}
```
at the top of the function and get rid of the `else` branch.

Reply:
Updated as requested.